### ISSUE TITLE
Abspath leaves file handles open

### ISSUE THREAD

**Issue by essenciary** [JuliaLang/julia#31126]
While doing some performance testing for Genie, running on 1000 requests with 50 concurrent ones, the app crashed due to the following error: 
```julia
Distributed.RemoteException(1, 
CapturedException(
Base.IOError("getcwd: too many open files (EMFILE)", -24), 
Any[(pwd() at sys.dylib:?, 1), 
(abspath at path.jl:325 [inlined], 1), 
(#file_path#40(::Bool, ::Function, ::String) at Router.jl:1065, 1), 
(file_path at Router.jl:1065 [inlined], 1), 
(is_static_file at Router.jl:952 [inlined], 1), 
(route_request(::Request, ::Response, ::IPv4) at Router.jl:89, 1), 

... etc

(handle_request(::Request, ::Response, ::IPv4) at AppServer.jl:96, 1), 
(handle_request(::Request, ::Response) at AppServer.jl:95, 1), 
(#14 at macros.jl:108 [inlined], 1), 
(run_work_thunk(::##14#17{Request,Response}, ::Bool) at process_messages.jl:56, 1), (#remotecall_fetch#148(::Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Function, ::LocalProcess) at remotecall.jl:364, 1), (remotecall_fetch(::Function, ::LocalProcess) at remotecall.jl:364, 1), (#remotecall_fetch#152(::Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Function, ::Int64) at remotecall.jl:406, 1), (remotecall_fetch(::Function, ::Int64) at remotecall.jl:406, 1), (setup_http_handler(::Request, ::Response) at AppServer.jl:60, 1), (setup_http_handler at AppServer.jl:59 [inlined], 1), (#6 at AppServer.jl:40 [inlined], 1), (handle at Handlers.jl:248 [inlined], 1), (handle(::RequestHandlerFunction{##6#12}, ::Stream{Request,Transaction{TCPSocket}}) at Handlers.jl:271, 1), (#4 at Handlers.jl:339 [inlined], 1), (macro expansion at Servers.jl:352 [inlined], 1), ((::##13#14{##4#5{RequestHandlerFunction{##6#12}},Transaction{TCPSocket},Stream{Request,Transaction{TCPSocket}}})() at task.jl:259, 1)]))
```

```julia
julia> versioninfo()
Julia Version 1.1.0
Commit 80516ca202 (2019-01-21 21:24 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
```

Initial Discourse thread here: https://discourse.julialang.org/t/poor-performance-of-abspath/20853

**Comment by StefanKarpinski** [comment]
This may be a libuv bug since our `pwd` function is just a thin wrapper calling `uv_cwd`:

https://github.com/JuliaLang/julia/blob/4236774ccd0cd9c02f1a2a42/base/file.jl#L50-L55

**Comment by StefanKarpinski** [comment]
cc @vtjnash—do you happen to know if this might be the case?

**Comment by vtjnash** [comment]
Sounds like it might be an Apple libc bug, as that's not a legal error for it to return (although that's a pretty minor nit, since it's supposed to return EACCES)

TIL: on macos, `getcwd()` is nearly identical to `realpath(".")` (with some performance optimizations)

I don't think this is an issue with `abspath`, but I wouldn't be surprised if it's a problem somewhere inside Genie or one of it's dependencies. I would monitor the result of `lsof` on the process and see if something looks like it's getting stuck.

```julia
julia> GC.enable(false); for i = 1:10^5; abspath("."); end; GC.enable(true)
```

**Comment by vtjnash** [comment]
closing because no response and can't reproduce

**Comment by essenciary** [comment]
@vtjnash Sorry, I didn't see the updates. I tested and `lsof` does not show anything out of the ordinary with the code you provided. 

I can't reproduce the issue with `Genie` either but I don't know why. On one hand I'm on a different macOS version so maybe that fixed it. Also, I refactored my code removed the calls to `abspath` (turned out I could very well use relative paths and save on some function calls). 

```julia
julia> versioninfo()
Julia Version 1.1.1
Commit 55e36cc (2019-05-16 04:10 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin15.6.0)
  CPU: Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
```

Can't test on Darwin 14.5 anymore.
